feat: tx history

defispartan opened the pull request:
## General Changes

Tx history page, accessible from top right wallet menu.

Features table of transaction history for active market and market switcher.

## Developer Notes

Works using react-query to cache requests based on [userAddress, subgraphUrl], no packages added for graphql support, queries are made using native `fetch` with graphql queries exported as strings in `src/modules/history`.

---

## Author Checklist

Please ensure you, the author, have gone through this checklist to ensure there is an efficient workflow for the reviewers.

- [x]  The base branch is set to `main`
- [x]  The title is using [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) formatting
- [ ]  The Github issue has been linked to the PR in the Development section
- [x]  The General Changes section has been filled out
- [x]  Developer Notes have been added (optional)

**If the PR is ready for review:**

- [x]  The PR is in `Open` state and not in `Draft` mode
- [x]  The `Ready for Dev Review` label has been added

## Reviewer Checklist

Please ensure you, as the reviewer(s), have gone through this checklist to ensure that the code changes are ready to ship safely and to help mitigate any downstream issues that may occur.

- [ ]  End-to-end tests are passing without any errors
- [ ]  Code style generally follows existing patterns
- [ ]  Code changes do not significantly increase the application bundle size
- [ ]  If there are new 3rd-party packages, they do not introduce potential security threats
- [ ]  If there are new environment variables being added, they have been added to the `.env.example` file as well as the pertinant `.github/actions/*` files
- [ ]  There are no CI changes, or they have been approved by the DevOps and Engineering team(s)
- [ ]  Code changes have been quality checked in the ephemeral URL
- [ ]  QA verification has been completed
- [ ]  There are two or more approvals from the core team
- [ ]  Squash and merge has been checked

defispartan commented:
Link T-8223

WOAftp reviewed (approved):
Io

WOAftp commented:
AMA

iamanastasia commented:
Noticed that it's hard to find assets by the name. E.g., there is WETH in the tx history, but I can't see it if I search for 'Ethereum'.
<img width="1521" alt="Screenshot 2023-05-24 at 15 59 03" src="https://github.com/aave/interface/assets/92739089/0ab83756-af9c-4185-9830-6d6ff6d103b0">

MareskoY commented:
Hello @defispartan, there are a few bugs that need to be addressed:

1. There seems to be a problem with the asset naming on the AMM market, for example, "AmmBptBALWETH". This needs to be tested on "aavechan.Ethan".
<img width="788" alt="Screenshot 2023-05-24 at 13 00 41" src="https://github.com/aave/interface/assets/18032275/80f322d2-4022-4b9d-a1ad-14782737a482">

2. When all types of transactions are chosen manually, the filter should show "All transactions".
<img width="620" alt="Screenshot 2023-05-24 at 13 03 46" src="https://github.com/aave/interface/assets/18032275/916d1296-e7f8-4daa-9d1f-7be870b7cb0a">

3. Resetting filters doesn't skip the search input, but it does skip it from the search results.
![filterbug](https://github.com/aave/interface/assets/18032275/50680388-3397-47b1-a699-5b5177513033)

4. Regarding CSV and JSON:
4.1. The amount is showing without decimals, but this may be okay.
4.2. In CSV, some amounts are shown from the right side and others from the left side.
4.3. The last three columns don't have a title.

5. The "View on Block Explorer" button is only clickable when under "Put," but not in the middle or at the top.

6. The "Upload" button looks strange, maybe we just need to scroll without the button.
<img width="847" alt="Screenshot 2023-05-24 at 13 34 09" src="https://github.com/aave/interface/assets/18032275/ef436681-5f8e-4db4-8515-dd8d3580796f">

defispartan commented:
@iamanastasia :  Added the ability to search by asset name as well as symbol.

@MareskoY :
1.) Fixed all AMM symbols to match what is shown on market/dashboard/reserve screens.
2.) Selecting all filters will now clear the rows and switch to "All Transactions". I wasn't sure if it was better to clear the rows or keep them all selected, maybe @iamanstasia can weigh in here?
3.) Search is cleared on reset now.
4.1.) Amounts are now formatted to the correct decimals on download.
4.2.) There is no right or left formatting in the csv, this is done by the program viewing the csv, usually based on the type of the column (number, string, etc.).
4.3) This was a bug where the "reserve" field was split into multiple columns. All rows should have the same # of columns now.
5.) Didn't understand this one. The arrow button is the only clickable part for this, the tooltip is not clickable.
6.) I removed the button border and it just shows a loading spinner at the bottom if it's fetching the next page or txns, or if it's fetching all data when you activate a filter (search or action filter)

defispartan commented:
@MareskoY That icon should be fixed now

foodaka commented:
@defispartan can you update the conflicts

grothem commented on src/modules/history/TransactionRowItem.tsx:
```diff
@@ -0,0 +1,157 @@
+  const theme = useTheme();
+
+  const downToMD = useMediaQuery(theme.breakpoints.down('md'));
+  const downToSM = useMediaQuery(theme.breakpoints.down('sm'));
```
And then could also remove `downToSM`

defispartan replied:
Removed

grothem commented on src/modules/history/actions/BorrowRateModeBlock.tsx:
```diff
@@ -0,0 +1,44 @@
+  swapBorrowRateTx: TransactionHistoryItem<ActionFields['SwapBorrowRate']>;
+  from?: boolean;
+}) => {
+  if (from && swapBorrowRateTx.borrowRateModeFrom === 'Variable') {
```
I think the `borrowRateModeFrom` and `borrowRateModeFrom` come back as the numeric values, so this should be updated to be
```
if (from && swapBorrowRateTx.borrowRateModeFrom === 2) {
...
}
```

defispartan replied:
I double checked and it does come back as a string, here's a sample swapBorrowRateTx: 

```
{
    "id": "11374100:19:0x62d0bbeab1336e25a593f49e53be8ec988ae97d20ccd06ec06b52b0feab0608d:37:37",
    "timestamp": 1606925645,
    "txHash": "0x62d0bbeab1336e25a593f49e53be8ec988ae97d20ccd06ec06b52b0feab0608d",
    "action": "Swap",
    "borrowRateModeFrom": "Variable",
    "borrowRateModeTo": "Stable",
    "variableBorrowRate": "33333347828159348205473884",
    "stableBorrowRate": "56666673914079674102736942",
    "reserve": {
        "symbol": "DAI",
        "decimals": 18,
        "name": "Dai Stablecoin",
        "underlyingAsset": "0x6b175474e89094c44da98b954eedeac495271d0f"
    }
}
```

grothem replied:
Ohh this is different between v2 and v3, I was trying on Optimism.

v2 shows `Variable` and `Stable` and v3 will have `2` and `1`

grothem commented on src/hooks/useTransactionHistory.tsx:
```diff
@@ -0,0 +1,245 @@
+      variables: { userAddress: account, first, skip, pool },
+    };
+    try {
+      const response = await fetch(subgraphUrl, {
```
Not a huge deal, but if a market doesn't have a subgraphUrl, it just shows the loading skeleton indefinitely. Only an issue for Scroll and Sepolia test nets right now. Maybe we could just show some message that history is not available if there's no subgraph url configured?

defispartan replied:
Good point, and that actually reminded me that all of the V3 testnet subgraphs needed to be disabled because the subgraphs do not match UI markets which were re-deployed w/ the permissioned faucet update. I added a placeholder for these markets:

![TxHistoryUnavailable](https://github.com/aave/interface/assets/26660211/8170ae6f-5ba7-4ce5-989d-ede6057da64c)

defispartan commented on src/modules/history/TransactionRowItem.tsx:
```diff
@@ -0,0 +1,157 @@
+
+interface TransactionHistoryItemProps {
+  transaction: TransactionHistoryItem & ActionFields[keyof ActionFields];
+  downToXSM: boolean;
```
Removed

defispartan commented on src/modules/history/HistorySearchBox.tsx:
Removed

JoaquinBattilana commented on src/modules/history/HistoryFilterMenu.tsx:
```diff
@@ -0,0 +1,258 @@
+    case FilterOptions.COLLATERALCHANGE:
+      return <Trans>Collateral change</Trans>;
+    case FilterOptions.LIQUIDATION:
+      return <Trans>Liqudation</Trans>;
```
Typo here, i don't add a code suggestion because it needs building for the translations

defispartan commented on src/modules/history/HistorySearchBox.tsx:
```diff
@@ -0,0 +1,58 @@
+      }}
+      value={searchQuery}
+      onChange={handleInputChange}
+      placeholder={'Search assets...'} // TODO: i18n
```
This component was removed, but it's also present in the component that replaced it: SearchInput. The issue is that the placeholder prop in InputBase is a string so we don't have an easy way to wrap it in a Trans tag

defispartan commented on src/modules/history/HistoryFilterMenu.tsx:
```diff
@@ -0,0 +1,258 @@
+
+const FilterLabel: React.FC<FilterLabelProps> = ({ filter }) => {
+  switch (filter) {
+    case FilterOptions.SUPPLY:
```
I have all re-used types in src/modules/history/types.tsx, this was the only place where actions were mapped to text directly so I moved it here to save an export

defispartan commented on src/layouts/WalletWidget.tsx:
```diff
@@ -193,7 +195,34 @@ export default function WalletWidget({ open, setOpen, headerHeight }: WalletWidg
       <Box
         component={component}
-        sx={{ color: { xs: '#F1F1F3', md: 'text.primary' } }}
+        sx={{ color: { xs: '#F1F1F3', md: 'text.primary', cursor: 'pointer' } }}
```
tbh there's a lot of places that could be moved to theme that should probably be cleaned up in the future 😅

defispartan commented on src/modules/history/HistoryFilterMenu.tsx:
```diff
@@ -0,0 +1,258 @@
+    onFilterChange(currentFilter);
+  };
+
+  const handleFilterClick = (filter: FilterOptions | undefined) => {
```
This would break when selecting "All transactions", the return statement would mean setLocalFilter would never get called with the empty filter array

defispartan commented on src/modules/history/HistoryWrapper.tsx:
```diff
@@ -0,0 +1,287 @@
+      csvContent += row.join(',') + '\n';
+    });
+
+    downloadData('transactions.csv', csvContent, 'text/csv');
```
Makes sense, it should be extremely rare since the only part that can fail is the `fetchForDownload` but I'll add it to a ticket of potential future changes
